### PR TITLE
Aircall Connector 

### LINKED ISSUE #1297: Aircall

### PR DESCRIPTION
**Pull request by tmuralikrishnan**
Closes #1297

### Code Changes
NEW SaaS Connector Aircall for access 

### Steps to Confirm
* [x] Add/configure the Aircall connector in UI
* [x] Verify Aircall config and icon UI looks good 
* [x] Add customer data using UI
* [x] Add calls data using Aircall App(no api for create calls)
* [x] Create an Access DSR with the new Aircall customer phone number 
* [x] Verify email sent by fides with customer data is correct 
* [x] Create an Delete DSR with the new Aircall customer phone number
* [x] Verify the customer data is now deleted in Aircall

### Pre-Merge Checklist
* [ ] All CI Pipelines Succeeded
* [ ] Update `CHANGELOG.md`

### SaaS Connector Checklist 
* [x] [Spec doc link](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2615377921/AirCall+Connector+Specification)
* [x] Vault Creds
* [x] Postman Collection
* [x] SaaS Config And Dataset 
* [x] Access and Erasure Test Passing 
* [ ] Icon

### CODE REVIEW

**Comment by tmuralikrishnan** [comment]
image

**Comment by tmuralikrishnan** [comment]
image

**Comment by tmuralikrishnan** [comment]
image

File: data/saas/config/aircall_config.yml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        headers:
+          - name: Authorization
+            value: Basic <api_token>
+
+  test_request:
+    method: GET
+    path: /v1/contacts/
+    param_values:
+      - name: api_token
+        connector_param: api_token
```
Comment (Kelsey-Ethyca) [inline data/saas/config/aircall_config.yml:28]: remove api_token from all calls now that it is in the client_config

File: data/saas/config/aircall_config.yml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+               <masked_object_fields> 
+            }
+          param_values:
+            - name: contact_id
+              references:
+                - dataset: <instance_fides_key>
+                  field: contact.id
+                  direction: from
+            - name: api_token
+              connector_param: api_token
```
Comment (Kelsey-Ethyca) [inline data/saas/config/aircall_config.yml:70]: remove update. We can only have an update or delete per endpoint.

File: data/saas/dataset/aircall_dataset.yml [hunk 3]
```diff
@@ -0,0 +1,413 @@
+dataset:
+  - fides_key: <instance_fides_key>
+    name: aircall
+    description: A sample dataset representing the Aircall connector for Fides
+    collections:
+      - name: customer
+        fields:
+          - name: id
+            data_categories: [system.operations]
```
Comment (Kelsey-Ethyca) [inline data/saas/dataset/aircall_dataset.yml:9]: this should be user.unique_id 

use this for reference https://ethyca.github.io/fideslang/explorer/

File: data/saas/dataset/aircall_dataset.yml [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+            fidesops_meta:
+              data_type: integer
+          - name: numbers
+            fidesops_meta:
+              data_type: 'object[]'
+            fields:
+              - name: '0'
+                fidesops_meta:
+                  data_type: object
```
Comment (Kelsey-Ethyca) [inline data/saas/dataset/aircall_dataset.yml:54]: un-nest and delete '0'

File: data/saas/dataset/aircall_dataset.yml [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+                fields:
+                  - name: id
+                    data_categories: [system.operations]
```
Comment (Kelsey-Ethyca) [inline data/saas/dataset/aircall_dataset.yml:57]: user.unique_id

File: tests/ops/integration_tests/saas/test_aircall_task.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            "teams",
+            "number",
+            "cost",
+            "country_code_a2",
+            "pricing_type",
+            "comments"
+        ],
+    )
+    # verify we only returned data for our identity email
+    assert v[f"{dataset_name}:contact"][0]["emails"][0]["value"] == aircall_identity_email
```
Comment (Kelsey-Ethyca) [inline tests/ops/integration_tests/saas/test_aircall_task.py:124]: add erasure tests (and fixtures) for contacts and calls

File: data/saas/config/aircall_config.yml (hunk 1 above)
Comment (tmuralikrishnan) [reply data/saas/config/aircall_config.yml]: api_token removed from all end points

File: data/saas/config/aircall_config.yml (hunk 2 above)
Comment (tmuralikrishnan) [reply data/saas/config/aircall_config.yml]: now removed update api

File: data/saas/dataset/aircall_dataset.yml (hunk 4 above)
Comment (tmuralikrishnan) [reply data/saas/dataset/aircall_dataset.yml]: nested fields with '0' removed

File: data/saas/dataset/aircall_dataset.yml (hunk 5 above)
Comment (tmuralikrishnan) [reply data/saas/dataset/aircall_dataset.yml]: data categories added

File: tests/ops/integration_tests/saas/test_aircall_task.py (hunk 6 above)
Comment (tmuralikrishnan) [reply tests/ops/integration_tests/saas/test_aircall_task.py]: Erasure request added.but,there is no api for create calls.Only possible to make a  call via Aircall App.So,now only checked contact deletion.

File: data/saas/dataset/aircall_dataset.yml (hunk 3 above)
Comment (tmuralikrishnan) [reply data/saas/dataset/aircall_dataset.yml]: data categories added

**Comment by tmuralikrishnan** [comment]
In user privacy request creation page phone number is required and email is optional for aircall.

File: data/saas/config/aircall_config.yml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  fides_key: <instance_fides_key>
+  name: Aircall SaaS Config
+  type: aircall
+  description: A sample schema representing the Aircall connector for Fides
+  version: 0.1.0
+
+  connector_params:
+    - name: domain
+      default_value: api.aircall.io
+    - name: api_token
```
Comment (galvana) [inline data/saas/config/aircall_config.yml:11]: This connector uses `basic` authentication so you should ask the user for `api_id` and `api_token`, these are used as the username and password.

```
  connector_params:
    - name: domain
      default_value: api.aircall.io
    - name: api_id
      label: API ID
    - name api_token
      label: API Token

  client_config:
    protocol: https
    host: <domain>
    authentication:
      strategy: basic
      configuration:
          - username: <api_id>
          - password: <api_token>
```

File: src/fides/data/test_env/fides.test_env.toml [hunk 8]
```diff
@@ -29,7 +29,7 @@ task_retry_delay = 1
 task_retry_backoff = 1
 require_manual_request_approval = true
 subject_identity_verification_required = false
-masking_strict = true
+masking_strict = false
```
Comment (galvana) [inline src/fides/data/test_env/fides.test_env.toml:32]: Please revert this change

File: src/fides/data/test_env/privacy_center_config/config.json [hunk 9]
```diff
@@ -12,8 +12,8 @@
       "description": "We will provide you a report of all your personal data.",
       "identity_inputs": {
         "name": "optional",
-        "email": "required",
```
Comment (galvana) [inline src/fides/data/test_env/privacy_center_config/config.json:15]: Please revert the changes in this file too

File: tests/fixtures/saas/aircall_fixtures.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+from tests.ops.test_helpers.vault_client import get_secrets
+
+secrets = get_secrets("aircall")
+
+
+@pytest.fixture(scope="session")
+def aircall_secrets(saas_config):
+    return {
+        "domain": pydash.get(saas_config, "aircall.domain") or secrets["domain"],        
+        "api_token": pydash.get(saas_config, "aircall.api_token") or secrets["api_token"],
```
Comment (galvana) [inline tests/fixtures/saas/aircall_fixtures.py:30]: Update this with `app_id` and `api_token`

File: tests/ops/integration_tests/saas/test_aircall_task.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+            "teams",
+            "number",
+            "cost",
+            "country_code_a2",
+            "pricing_type",
+            "comments"
+        ],
+    )
+    # verify we only returned data for our identity email
+    assert v[f"{dataset_name}:contact"][0]["phone_numbers"][0]["value"] == aircall_identity_phone_number
```
Comment (galvana) [inline tests/ops/integration_tests/saas/test_aircall_task.py:106]: You need to check the `calls` as well, not just the `contacts`

File: tests/ops/integration_tests/saas/test_aircall_task.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    db,
+    policy,
+    erasure_policy_string_rewrite,
+    aircall_connection_config,
+    aircall_dataset_config,
+    aircall_erasure_identity_phone_number,
+    aircall_erasure_identity_email,
+    aircall_create_erasure_data,
+) -> None:
+    """Full erasure request based on the aircall SaaS config"""
```
Comment (galvana) [inline tests/ops/integration_tests/saas/test_aircall_task.py:122]: nit: Capitalize `Aircall`

File: tests/ops/integration_tests/saas/test_aircall_task.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+            "is_shared",
+            "direct_link",
+            "created_at",
+            "updated_at",
+            "phone_numbers",
+            "emails",
+        ],
+    )
+
+    # assert_rows_match(
```
Comment (galvana) [inline tests/ops/integration_tests/saas/test_aircall_task.py:164]: Don't leave commented code in the final version

**Review by galvana — CHANGES_REQUESTED** [review]
Overall this looks good but we need to update the authentication strategy to `basic`. Also be mindful of not making changes to code that is not related to this connector (like fides.toml and the privacy center config) and leaving commented code

File: tests/fixtures/saas/aircall_fixtures.py (hunk 10 above)
Comment (soundayra21) [reply tests/fixtures/saas/aircall_fixtures.py:32]: updated app_id and api_token in aircall_fixtures.py

File: tests/ops/integration_tests/saas/test_aircall_task.py (hunk 11 above)
Comment (soundayra21) [reply tests/ops/integration_tests/saas/test_aircall_task.py:105]: Checked the calls endpoint as well

File: tests/ops/integration_tests/saas/test_aircall_task.py (hunk 12 above)
Comment (soundayra21) [reply tests/ops/integration_tests/saas/test_aircall_task.py]: Capitalized the Aircall text

File: tests/ops/integration_tests/saas/test_aircall_task.py (hunk 13 above)
Comment (soundayra21) [reply tests/ops/integration_tests/saas/test_aircall_task.py]: removed the commented code

File: tests/ops/integration_tests/saas/test_aircall_task.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+        erasure_policy_string_rewrite,
+        graph,
+        [aircall_connection_config],
+        {"phone_number": aircall_erasure_identity_phone_number},
+        get_cached_data_for_erasures(privacy_request.id),
+        db,
+    )
+
+    assert x == {
+        # f"{dataset_name}:customer": 0,
```
[the remark this answers is not in the document]
Comment (soundayra21) [reply tests/ops/integration_tests/saas/test_aircall_task.py]: removed the commented code

**Review by galvana — APPROVED** [review]
